Feat: Use oracle for token price - added support for Pyth Orace

0xCipherCoder opened the pull request:
This PR adds Pyth Oracle to get a real-time price from the Pyth network. Added test case and retry mechanism for the changes made.
. 
 This pull request was created for https://app.gib.work/bounties/9429edb4-175f-40c5-bc29-eb57e79b3a71 in an attempt to solve a bounty #12 . Payment for the bounty is immediately sent to the contributor after merge.

0xCipherCoder commented:
I tried using Switchboard as well to have combined price but there are multiple dependency issues with their Solana SDK.

gitteri commented on crates/lib/src/oracle/pyth.rs:
```diff
@@ -0,0 +1,173 @@
+const RETRY_DELAY_MS: u64 = 1000;
+
+lazy_static! {
+    static ref PYTH_PRICE_ACCOUNTS: HashMap<&'static str, &'static str> = {
+        let mut m = HashMap::new();
+        // SOL/USD
+        m.insert(
+            "So11111111111111111111111111111111111111112",
+            "H6ARHf6YXhGYeQfUzQNGk6rDNnLBQKrenN712K4AQJEG"
+        );
+        // USDC/USD
+        m.insert(
+            "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
+            "Gnt27xtC473ZT2Mw5u8wZ68Z3gULkSTb5DuxJy7eJotD"
+        );
+        // USDT/USD
+        m.insert(
+            "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB",
+            "3vxLXJqLqF3JG5TCbYycbKWRBbCJQLxQmBGCkyqEEefL"
+        );
+        m
+    };
```
Is there a way to get the correct price accounts dynamically instead of hard coding these?

0xCipherCoder replied:
I am checking this

0xCipherCoder replied:
@gitteri I have added functionality for getting all USDC pairs supported by Pyth and checking for whitelisted token symbols with mint stored in the config. This will help support any token that is whitelisted in the future. Let me know if the approach is fine.

gitteri commented on crates/lib/src/transaction/paid_transaction.rs:
```diff
@@ -26,14 +27,24 @@ pub async fn sign_transaction_if_paid(
+        .get_token_price(&sol_mint)
+        .await
+        .map(|price| TokenPriceInfo { price })
+        .unwrap_or(TokenPriceInfo { price: 0.0 });
```
Instead of a 0 price we should return an error

0xCipherCoder replied:
Sure will do.

0xCipherCoder replied:
I am throwing an error now in case of price is not available from Oracle.

0xCipherCoder commented:
@gitteri I added and updated test cases wherever changes have been made. Please find the screenshots for your reference.
![Screenshot 2025-02-14 103359](https://github.com/user-attachments/assets/b391e3bd-aeaa-47fc-81b2-5de7942a1a6f)
![Screenshot 2025-02-14 103315](https://github.com/user-attachments/assets/78393056-e384-4263-b000-4ad630d0fa61)

gitteri commented on crates/lib/src/config.rs:
```diff
@@ -57,16 +58,17 @@ mod tests {
+        allowed_tokens = ["token1", "token2"]
+        allowed_spl_paid_tokens = ["token3"]
+        disallowed_accounts = ["account1"]
+        token_symbols = { "So11111111111111111111111111111111111111112" = "SOL", "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v" = "USDC", "Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB" = "USDT" }
```
You should be able to get the symbol from the token itself instead of relying on manual mapping
